File: packages/android-kotlin/src/main/kotlin/com/cinacoin/appkit/auth/AuthManager.kt
package com.cinacoin.appkit.auth

import android.content.Context
import com.cinacoin.appkit.auth.model.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Unified auth manager combining social and email login.
 * Handles token refresh, session persistence, and auto-refresh scheduling.
 */
object AuthManager {

    private val _currentUser = MutableStateFlow<AuthResult?>(null)
    val currentUser: StateFlow<AuthResult?> = _currentUser

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated

    private var socialLogin: SocialLoginManager? = null
    private var emailLogin: EmailLoginManager? = null

    private var authUrl: String = "https://auth.cinacoin.com"
    private var projectId: String = ""

    /** Stored application context for auto-refresh (weak reference to avoid leaks) */
    private var appContext: android.content.Context? = null

    /** How many seconds before expiry to trigger a refresh */
    private const val REFRESH_THRESHOLD_MS = 5 * 60 * 1000L // 5 minutes

    /** Job for the auto-refresh timer */
    private var refreshJob: Job? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    fun configure(
        context: Context,
        authUrl: String = "https://auth.cinacoin.com",
        projectId: String
    ) {
        this.appContext = context.applicationContext
        this.authUrl = authUrl
        this.projectId = projectId
        socialLogin = SocialLoginManager(authUrl, projectId)
        emailLogin = EmailLoginManager(authUrl, projectId)

        // Restore session from secure storage
        restoreSession(context)
    }

    /**
     * Restore session from secure storage.
     * Must be called with an Android Context (e.g., Application context).
     */
    fun restoreSession(context: Context) {
        val prefs = SecureTokenStorage.getPreferences(context)
        val accessToken = prefs.getString(SecureTokenStorage.KEY_ACCESS_TOKEN, null) ?: return
        val userId = prefs.getString(SecureTokenStorage.KEY_USER_ID, null) ?: return
        val expiresAt = prefs.getLong(SecureTokenStorage.KEY_EXPIRES_AT, 0)
        val refreshToken = prefs.getString(SecureTokenStorage.KEY_REFRESH_TOKEN, null)

        // Don't restore if token is already expired
        if (System.currentTimeMillis() >= expiresAt) {
            signOut(context)
            return
        }

        val restored = AuthResult(
            userId = userId,
            email = null,
            provider = null,
            accessToken = accessToken,
            refreshToken = refreshToken,
            expiresAt = expiresAt
        )

        _currentUser.value = restored
        _isAuthenticated.value = true
        scheduleAutoRefresh(restored)
    }

    // MARK: - Social Login

    suspend fun signInWithGoogle(context: Context): AuthResult {
        val manager = socialLogin ?: throw AuthError.NotConfigured
        val result = manager.signInWithGoogle(context)
        handleAuthResult(context, result)
        return result
    }

    suspend fun signInWithGitHub(context: Context): AuthResult {
        val manager = socialLogin ?: throw AuthError.NotConfigured
        val result = manager.signInWithGitHub(context)
        handleAuthResult(context, result)
        return result
    }

    suspend fun signInWithDiscord(context: Context): AuthResult {
        val manager = socialLogin ?: throw AuthError.NotConfigured
        val result = manager.signInWithDiscord(context)
        handleAuthResult(context, result)
        return result
    }

    // MARK: - Email Login

    suspend fun register(email: String, password: String): AuthResult {
        val manager = emailLogin ?: throw AuthError.NotConfigured
        val result = manager.register(email, password)
        return result
    }

    suspend fun login(email: String, password: String): AuthResult {
        val manager = emailLogin ?: throw AuthError.NotConfigured
        val result = manager.login(email, password)
        return result
    }

    // MARK: - Session

    fun signOut(context: Context? = null) {
        _currentUser.value = null
        _isAuthenticated.value = false
        cancelRefreshTimer()

        context?.let { SecureTokenStorage.clear(it) }
    }

    // MARK: - Token Refresh

    /**
     * Refresh the access token using the stored refresh token.
     * Handles refresh token rotation: if the server returns a new refresh token, it is stored.
     */
    suspend fun refreshToken(context: Context): AuthResult = withContext(Dispatchers.IO) {
        val current = _currentUser.value ?: throw AuthError.TokenExpired
        val currentRefreshToken = current.refreshToken ?: throw AuthError.TokenExpired

        val url = URL("$authUrl/api/auth/oauth/refresh")
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/json")
        conn.doOutput = true

        val body = JSONObject().apply {
            put("refresh_token", currentRefreshToken)
            put("project_id", projectId)
        }
        conn.outputStream.write(body.toString().toByteArray())

        val responseCode = conn.responseCode
        if (responseCode !in 200..299) {
            val errorBody = try {
                conn.errorStream?.bufferedReader()?.readText() ?: "Unknown error"
            } catch (_: Exception) {
                "Unknown error"
            }
            throw AuthError.NetworkError("Token refresh failed: $responseCode - $errorBody")
        }

        val response = conn.inputStream.bufferedReader().readText()
        val obj = JSONObject(response)

        val newAccessToken = obj.getString("access_token")
        val userId = obj.getString("user_id")
        val expiresAt = obj.optLong("expires_at", 0)
        // Handle refresh token rotation
        val newRefreshToken = obj.optString("refresh_token", null) ?: currentRefreshToken
        val email = obj.optString("email", null)

        val result = AuthResult(
            userId = userId,
            email = email,
            provider = current.provider,
            accessToken = newAccessToken,
            refreshToken = newRefreshToken,
            expiresAt = expiresAt
        )

        // Persist new tokens
        handleAuthResult(context, result)

        result
    }

    /**
     * Ensure the current token is valid, refreshing if needed.
     * Returns a valid access token string.
     */
    suspend fun ensureValidToken(context: Context): String {
        val current = _currentUser.value ?: throw AuthError.TokenExpired

        val timeUntilExpiry = current.expiresAt - System.currentTimeMillis()
        if (timeUntilExpiry > REFRESH_THRESHOLD_MS) {
            return current.accessToken
        }

        // Token is about to expire or already expired — refresh
        val refreshed = refreshToken(context)
        return refreshed.accessToken
    }

    // MARK: - Callback Handling

    /**
     * Handle an incoming deep link intent.
     * Call this from your Activity's onNewIntent or onCreate when handling app links.
     * Returns true if the intent was an auth callback that was handled.
     */
    fun handleCallbackIntent(intent: android.content.Intent?): Boolean {
        val manager = socialLogin ?: return false
        if (!manager.isAuthCallback(intent)) return false
        manager.handleCallbackIntent(intent)
        return true
    }

    // MARK: - Private

    private fun handleAuthResult(context: Context, result: AuthResult) {
        _currentUser.value = result
        _isAuthenticated.value = true

        // Persist tokens
        val prefs = SecureTokenStorage.getPreferences(context)
        prefs.edit().apply {
            putString(SecureTokenStorage.KEY_ACCESS_TOKEN, result.accessToken)
            result.refreshToken?.let {
                putString(SecureTokenStorage.KEY_REFRESH_TOKEN, it)
            }
            putString(SecureTokenStorage.KEY_USER_ID, result.userId)
            putLong(SecureTokenStorage.KEY_EXPIRES_AT, result.expiresAt)
            apply()
        }

        // Schedule auto-refresh
        scheduleAutoRefresh(result)
    }

    /**
     * Schedule automatic token refresh before expiry.
     */
    private fun scheduleAutoRefresh(result: AuthResult) {
        cancelRefreshTimer()

        val ctx = appContext ?: return
        val timeUntilExpiry = result.expiresAt - System.currentTimeMillis()
        val refreshIn = maxOf(timeUntilExpiry - REFRESH_THRESHOLD_MS, 10_000L)

        refreshJob = scope.launch {
            delay(refreshIn)
            try {
                val current = _currentUser.value ?: return@launch
                if (current.refreshToken == null) return@launch

                println("[AuthManager] Auto-refresh triggered")
                refreshToken(ctx)
            } catch (e: Exception) {
                println("[AuthManager] Auto-refresh failed: ${e.message}")
                scheduleRetryRefresh()
            }
        }
    }

    private fun scheduleRetryRefresh() {
        cancelRefreshTimer()
        val ctx = appContext ?: return
        refreshJob = scope.launch {
            delay(60_000L)
            try {
                println("[AuthManager] Retry refresh triggered")
                refreshToken(ctx)
            } catch (e: Exception) {
                println("[AuthManager] Retry refresh failed: ${e.message}")
            }
        }
    }

    private fun cancelRefreshTimer() {
        refreshJob?.cancel()
        refreshJob = null
    }
}

File: packages/ios-swift/Sources/CinacoinAppKitAuth/AuthManager.swift
import Foundation

/// Unified auth manager combining social and email login
public final class AuthManager: ObservableObject, @unchecked Sendable {
    
    public static let shared = AuthManager()
    
    @Published public private(set) var currentUser: AuthResult?
    @Published public private(set) var isAuthenticated = false
    
    private var socialLogin: SocialLoginManager?
    private var emailLogin: EmailLoginManager?
    
    private let authUrl: String
    private let projectId: String
    
    /// Timer for auto-refresh
    private var refreshTimer: Timer?
    
    /// How many seconds before expiry to trigger a refresh
    private let refreshThresholdSeconds: TimeInterval = 300 // 5 minutes
    
    private init() {
        self.authUrl = "https://auth.cinacoin.com"
        self.projectId = ""
    }
    
    /// Configure auth manager
    public func configure(authUrl: String = "https://auth.cinacoin.com", projectId: String) {
        socialLogin = SocialLoginManager(authUrl: authUrl, projectId: projectId)
        emailLogin = EmailLoginManager(authUrl: authUrl, projectId: projectId)
        
        // Restore session from keychain
        restoreSession()
    }
    
    // MARK: - Social Login
    
    public func signInWithApple() async throws -> AuthResult {
        guard let socialLogin = socialLogin else {
            throw AuthError.notImplemented("Auth not configured. Call configure() first.")
        }
        let result = try await socialLogin.signInWithApple()
        await handleAuthResult(result)
        return result
    }
    
    public func signInWithGoogle() async throws -> AuthResult {
        guard let socialLogin = socialLogin else {
            throw AuthError.notImplemented("Auth not configured")
        }
        let result = try await socialLogin.signInWithGoogle()
        await handleAuthResult(result)
        return result
    }
    
    public func signInWithGitHub() async throws -> AuthResult {
        guard let socialLogin = socialLogin else {
            throw AuthError.notImplemented("Auth not configured")
        }
        let result = try await socialLogin.signInWithGitHub()
        await handleAuthResult(result)
        return result
    }
    
    public func signInWithDiscord() async throws -> AuthResult {
        guard let socialLogin = socialLogin else {
            throw AuthError.notImplemented("Auth not configured")
        }
        let result = try await socialLogin.signInWithDiscord()
        await handleAuthResult(result)
        return result
    }
    
    // MARK: - Email Login
    
    public func register(email: String, password: String) async throws -> AuthResult {
        guard let emailLogin = emailLogin else {
            throw AuthError.notImplemented("Auth not configured")
        }
        let result = try await emailLogin.register(email: email, password: password)
        await handleAuthResult(result)
        return result
    }
    
    public func login(email: String, password: String) async throws -> AuthResult {
        guard let emailLogin = emailLogin else {
            throw AuthError.notImplemented("Auth not configured")
        }
        let result = try await emailLogin.login(email: email, password: password)
        await handleAuthResult(result)
        return result
    }
    
    // MARK: - Session Management
    
    public func signOut() {
        currentUser = nil
        isAuthenticated = false
        cancelRefreshTimer()
        
        // Remove all tokens from keychain
        try? KeychainHelper.delete(key: KeychainKeys.accessToken)
        try? KeychainHelper.delete(key: KeychainKeys.refreshToken)
        try? KeychainHelper.delete(key: KeychainKeys.userId)
        try? KeychainHelper.delete(key: KeychainKeys.expiresAt)
    }
    
    // MARK: - Token Refresh
    
    /// Refresh the access token using the stored refresh token
    public func refreshToken() async throws -> AuthResult {
        guard let current = currentUser,
              let refreshTokenValue = current.refreshToken else {
            throw AuthError.tokenExpired
        }
        
        let url = URL(string: "\(authUrl)/api/auth/oauth/refresh")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "refresh_token": refreshTokenValue,
            "project_id": projectId
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            throw AuthError.refreshFailed("Server returned status \(statusCode)")
        }
        
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let accessToken = json["access_token"] as? String,
              let userId = json["user_id"] as? String,
              let expiresAt = json["expires_at"] as? TimeInterval else {
            throw AuthError.refreshFailed("Invalid response format")
        }
        
        let newRefreshToken = json["refresh_token"] as? String
        let email = json["email"] as? String
        
        // Handle refresh token rotation: if server returns a new refresh token, use it
        let finalRefreshToken = newRefreshToken ?? refreshTokenValue
        
        let result = AuthResult(
            userId: userId,
            email: email,
            provider: current.provider,
            accessToken: accessToken,
            refreshToken: finalRefreshToken,
            expiresAt: Date(timeIntervalSince1970: expiresAt)
        )
        
        // Update stored tokens
        await handleAuthResult(result)
        
        return result
    }
    
    /// Ensure the current token is valid, refreshing if needed
    public func ensureValidToken() async throws -> String {
        guard let current = currentUser else {
            throw AuthError.tokenExpired
        }
        
        // If token is still valid for at least `refreshThresholdSeconds`, return it
        let timeUntilExpiry = current.expiresAt.timeIntervalSinceNow
        if timeUntilExpiry > refreshThresholdSeconds {
            return current.accessToken
        }
        
        // Token is about to expire or already expired — refresh
        let refreshed = try await refreshToken()
        return refreshed.accessToken
    }
    
    // MARK: - Private
    
    @MainActor
    private func handleAuthResult(_ result: AuthResult) {
        currentUser = result
        isAuthenticated = true
        
        // Store in keychain
        if let tokenData = result.accessToken.data(using: .utf8) {
            try? KeychainHelper.save(key: KeychainKeys.accessToken, data: tokenData)
        }
        if let refreshData = result.refreshToken?.data(using: .utf8) {
            try? KeychainHelper.save(key: KeychainKeys.refreshToken, data: refreshData)
        }
        if let userIdData = result.userId.data(using: .utf8) {
            try? KeychainHelper.save(key: KeychainKeys.userId, data: userIdData)
        }
        let expiresAtTimestamp = String(result.expiresAt.timeIntervalSince1970)
        if let expiresData = expiresAtTimestamp.data(using: .utf8) {
            try? KeychainHelper.save(key: KeychainKeys.expiresAt, data: expiresData)
        }
        
        // Schedule auto-refresh
        scheduleAutoRefresh(for: result)
    }
    
    /// Schedule automatic token refresh before expiry
    private func scheduleAutoRefresh(for result: AuthResult) {
        cancelRefreshTimer()
        
        let timeUntilExpiry = result.expiresAt.timeIntervalSinceNow
        let refreshIn = max(timeUntilExpiry - refreshThresholdSeconds, 10) // At least 10 seconds from now
        
        refreshTimer = Timer.scheduledTimer(withTimeInterval: refreshIn, repeats: false) { [weak self] _ in
            Task { [weak self] in
                do {
                    _ = try await self?.refreshToken()
                } catch {
                    print("[AuthManager] Auto-refresh failed: \(error.localizedDescription)")
                    // Try again in 60 seconds
                    self?.scheduleRetryRefresh()
                }
            }
        }
    }
    
    /// Retry refresh after a failure
    private func scheduleRetryRefresh() {
        cancelRefreshTimer()
        refreshTimer = Timer.scheduledTimer(withTimeInterval: 60, repeats: false) { [weak self] _ in
            Task { [weak self] in
                do {
                    _ = try await self?.refreshToken()
                } catch {
                    print("[AuthManager] Retry refresh failed: \(error.localizedDescription)")
                }
            }
        }
    }
    
    private func cancelRefreshTimer() {
        refreshTimer?.invalidate()
        refreshTimer = nil
    }
    
    private func restoreSession() {
        guard let accessTokenData = try? KeychainHelper.load(key: KeychainKeys.accessToken),
              let accessToken = String(data: accessTokenData, encoding: .utf8),
              let userIdData = try? KeychainHelper.load(key: KeychainKeys.userId),
              let userId = String(data: userIdData, encoding: .utf8),
              let expiresAtData = try? KeychainHelper.load(key: KeychainKeys.expiresAt),
              let expiresAtStr = String(data: expiresAtData, encoding: .utf8),
              let expiresAtTimestamp = Double(expiresAtStr) else {
            return
        }
        
        let expiresAt = Date(timeIntervalSince1970: expiresAtTimestamp)
        
        // Don't restore if token is already expired
        guard expiresAt > Date() else {
            signOut()
            return
        }
        
        let refreshTokenValue: String? = {
            guard let data = try? KeychainHelper.load(key: KeychainKeys.refreshToken),
                  let str = String(data: data, encoding: .utf8) else { return nil }
            return str
        }()
        
        let restored = AuthResult(
            userId: userId,
            email: nil,
            provider: nil,
            accessToken: accessToken,
            refreshToken: refreshTokenValue,
            expiresAt: expiresAt
        )
        
        Task { @MainActor in
            self.currentUser = restored
            self.isAuthenticated = true
            self.scheduleAutoRefresh(for: restored)
        }
    }
}

// MARK: - Keychain Keys (shared with SocialLogin)
enum AuthKeychainKeys {
    static let accessToken = "com.cinacoin.auth.access_token"
    static let refreshToken = "com.cinacoin.auth.refresh_token"
    static let userId = "com.cinacoin.auth.user_id"
    static let expiresAt = "com.cinacoin.auth.expires_at"
}

/// Simple keychain helper
enum KeychainHelper {
    static func save(key: String, data: Data) throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock
        ]
        SecItemDelete(query as CFDictionary)
        let status = SecItemAdd(query as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw AuthError.networkError("Keychain save failed: \(status)")
        }
    }
    
    static func load(key: String) throws -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess || status == errSecItemNotFound else {
            throw AuthError.networkError("Keychain load failed: \(status)")
        }
        return result as? Data
    }
    
    static func delete(key: String) throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key
        ]
        SecItemDelete(query as CFDictionary)
    }
}
